### PR TITLE
fix DataLoaderDispatcher issue in Gaudi

### PR DESCRIPTION
**Pull request by sywangyi**
TRL SFT trainer has used IterableDataset and DataLoaderDispatcher will be called. The data shape/tensor will be broadcasted from rank 0 to other rank in DDP.

# What does this PR do?

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

### CODE REVIEW

**Comment by sywangyi** [comment]
Traceback (most recent call last):
  File "/intel-extension-for-transformers/optimum-habana/examples/trl_example/stack_llama_2/sft_llama2.py", line 184, in <module>
    trainer.train()
  File "/usr/local/lib/python3.10/dist-packages/trl/trainer/sft_trainer.py", line 281, in train
    output = super().train(*args, **kwargs)
  File "/intel-extension-for-transformers/optimum-habana/optimum/habana/transformers/trainer.py", line 486, in train
    return inner_training_loop(
  File "/intel-extension-for-transformers/optimum-habana/optimum/habana/transformers/trainer.py", line 807, in _inner_training_loop
    for step, inputs in enumerate(epoch_iterator):
  File "/usr/local/lib/python3.10/dist-packages/accelerate/data_loader.py", line 664, in __iter__
    raise ValueError(
ValueError: Batch does not contain any data (`None`). At the end of all iterable data available before expected stop iteration.
Traceback (most recent call last):
  File "/intel-extension-for-transformers/optimum-habana/examples/trl_example/stack_llama_2/sft_llama2.py", line 184, in <module>
    trainer.train()
  File "/usr/local/lib/python3.10/dist-packages/trl/trainer/sft_trainer.py", line 281, in train
    output = super().train(*args, **kwargs)
  File "/intel-extension-for-transformers/optimum-habana/optimum/habana/transformers/trainer.py", line 486, in train
    return inner_training_loop(
  File "/intel-extension-for-transformers/optimum-habana/optimum/habana/transformers/trainer.py", line 807, in _inner_training_loop
    for step, inputs in enumerate(epoch_iterator):
  File "/usr/local/lib/python3.10/dist-packages/accelerate/data_loader.py", line 664, in __iter__
    raise ValueError(
ValueError: Batch does not contain any data (`None`). At the end of all iterable data available before expected stop iteration.

File: optimum/habana/accelerate/data_loader.py [hunk 1]
```diff
@@ -1,18 +1,269 @@
...
+            The list of objects to broadcast. This list will be modified inplace.
+        from_process (`int`, *optional*, defaults to 0):
+            The process from which to send the data.
+
+    Returns:
+        The same list containing the objects from process 0.
+    """
+    if GaudiPartialState().distributed_type in [GaudiDistributedType.MULTI_HPU, GaudiDistributedType.DEEPSPEED]:
+        torch.distributed.broadcast_object_list(object_list, src=from_process, device="hpu")
+    return object_list
```
Comment (regisss) [inline optimum/habana/accelerate/data_loader.py:77]: Let's move these methods to `optimum/habana/accelerate/utils/operations.py` to keep the same tree structure as Accelerate. That makes maintenance easier when upgrading to new versions.

File: optimum/habana/accelerate/data_loader.py [hunk 2]
```diff
@@ -1,18 +1,269 @@
...
+def initialize_tensors(data_structure):
+    """
+    Recursively initializes tensors from a nested list/tuple/dictionary of [`~utils.TensorInformation`].
+
+    Returns:
+        The same data structure as `data` with tensors instead of [`~utils.TensorInformation`].
+    """
+
+    def _initialize_tensor(tensor_info):
+        return torch.zeros(*tensor_info.shape, dtype=tensor_info.dtype)
```
Comment (regisss) [inline optimum/habana/accelerate/data_loader.py:39]: I see that the latest release of Accelerate uses `torch.empty`: https://github.com/huggingface/accelerate/blob/v0.25.0/src/accelerate/utils/operations.py#L220
Any specific reason to use `torch.zeros` here?

File: optimum/habana/accelerate/data_loader.py [hunk 3]
```diff
@@ -1,18 +1,269 @@
...
+
+    Returns:
+        The same list containing the objects from process 0.
+    """
+    if GaudiPartialState().distributed_type in [GaudiDistributedType.MULTI_HPU, GaudiDistributedType.DEEPSPEED]:
+        torch.distributed.broadcast_object_list(object_list, src=from_process, device="hpu")
+    return object_list
+
+
+class GaudiDataLoaderDispatcher(DataLoader, DataLoaderStateMixin):
```
Comment (regisss) [inline optimum/habana/accelerate/data_loader.py:80]: The only changes are the used of the methods above and not the ones of Accelerate right?

File: optimum/habana/accelerate/data_loader.py (hunk 2 above)
Comment (sywangyi) [reply optimum/habana/accelerate/data_loader.py:39]: yes. meeting error like "Error when trying to cast Long to Int, Input values range [0, 72340172838010880] exceeds Int range [-2147483648, 2147483647]" if use torch.empty

File: optimum/habana/accelerate/data_loader.py (hunk 3 above)
Comment (sywangyi) [reply optimum/habana/accelerate/data_loader.py:80]: yes

File: optimum/habana/accelerate/data_loader.py (hunk 1 above)
Comment (sywangyi) [reply optimum/habana/accelerate/data_loader.py:77]: will do it.

**Review by regisss — APPROVED** [review]
LGTM

**Pull request merged**